### ISSUE TITLE
[API Proposal]: `ToString(...)` overloads for `Array`, `Array<T>`, `Memory<T>`, `Span<T>` for pretty printing underlying contents/elements and 

### ISSUE THREAD

**Issue by vadimkantorov** [dotnet/runtime#106531]
### Background and motivation

For code working with arrays/tensors/interop, this is extremely useful to have such a shortcut in the standard library.

In Python / NumPy, str impl for lists prints the full contents and the default impl for NumPy arrays - prints some leading and trailing elements.

I propose to add an overload for `ToString` method accepting some format string (and then we could specify if we want all elements or some leading/trailing only). Same for char arrays.

Essentially, getting C# closer to Python's `print(arr1, arr2)` would be great. Currently a novice has to search online of how to print an array which adds friction.

It could also be nice to introduce a new method `System.PrettyPrint(...)` or `System.Print(...)` which would by default recursively print structs/records/classes/arrays in a more pretty way (e.g. for classes/structs/records - listing all fields and their values) - and it could also be closer to Python's `print(...)` which can accept multiple values without passing a format string as the first arg.

### API Proposal

N/A

### API Usage

N/A

**Comment by julealgon** [comment]
The second part of this proposal is odd to me..... I don't think the framework should be concerned with that. Can't you just Json-serialize the model?

**Comment by vadimkantorov** [comment]
Of course, one can json-serialize or do `string.Join("\n", myarr)` or write some custom code, or accomplish the task in another way. My point is that for people working a lot with arrays (and coming from Go/Pyhon/scientific background or simply novices/students), it might be beneficial to have such shortcut in the standard library, so that printing contents of an array does not require googling / or loading a JSON library.

(And as my proposal is about adding a shortcut, of course its need is subjective, understanding that)

**Comment by Clockwork-Muse** [comment]
... part of the problem here is `ToString()` often ends up being a shortcut for two different behaviors: 
 - "Official"/pretty output
 - debug/full output

(this is explicitly spelled out with Rust, which might be nice here too...)

Generally, getting a string like this is mostly useful for debug output (for cases when you can't attach to a debugger) - for everything else, getting "correct" serialized output (or outputting to whatever stream directly) is better.  But I generally feel that debug output should be more targeted....

**Comment by vadimkantorov** [comment]
I was thinking of introducing an extra overload like `Array.ToString(string format)`. Then at least we can do `System.Console.WriteLine("{0:pretty}", array)`. This way the existing `.ToString()` behavior would be unchanged. And for recursive application - maybe a separate, new function would need to be introduced (like `System.Print(...)`) which would pass in this format modifier for some known types and maybe do some custom logic for classes if they have only a default not very pretty `ToString`)

**Comment by colejohnson66** [comment]
If you're debugging, why not just use the debugger's collection viewer?

**Comment by vadimkantorov** [comment]
> If you're debugging, why not just use the debugger's collection viewer?

I'm mainly using command-line and debug-printing.

Similar pretty printing would be extremely useful if C#/.NET starts shipping an official REPL tool.

**Comment by filipnavara** [comment]
> Similar pretty/debug printing would be extremely useful if C#/.NET starts shipping an official REPL tool.

Side note: There's already https://github.com/dotnet/interactive and https://github.com/jonsequitur/dotnet-repl built on top.

**Comment by vadimkantorov** [comment]
> There's already

Yeah, in my personal view it's time for a C# REPL/scripting be officially shipped with `dotnet` tool by default - for education and quick/laconic scripting - like it is with F#. Especially if the underlying foundation is already built and officially supported for this notebook endeavor.

**Comment by tannergooding** [comment]
Closing as this is currently inactionable as is. There is no proposed API surface and the "design work" is being left up to the team, which is not something we want to drive for every possible suggestion.

If you'd still like to see this happen, then please adjust the top comment (or open a new issue) that contains a full and complete API proposal, following the required template.

I'd note that `Tensor<T>` and `TensorSpan<T>` does provide this type of functionality. The default `ToString()` API only lists the tensor shape, not its contents. There is then a `ToString(params scoped ROSpan<...> maxLengths)` API that can be used to include the contents in a bounded way and which might serve as inspiration for how to provide this same functionality to array, span, list, or other collection types.
